Added support for PHP 8.1

driehle opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | yes
| QA            | no

### Description

This PR adds support for PHP 8.1. Besides, laminas-coding-standard has been updated to 2.0.

Note that I have suppressed a couple of type hint errors in `phpcs.xml`. Those rules are simply too noisy and I cannot fix all the issues there.

driehle committed:
updated dependencies to latest versions supporting PHP 7.3

Signed-off-by: Dennis Riehle <webmaster@riehle-web.com>

driehle committed:
allow PHP 8.1 and update laminas-coding-standard

Signed-off-by: Dennis Riehle <webmaster@riehle-web.com>

driehle committed:
fixed cs issues with phpcbf

Signed-off-by: Dennis Riehle <webmaster@riehle-web.com>

driehle committed:
fixed remaining issues, ignoring missing type hints

Signed-off-by: Dennis Riehle <webmaster@riehle-web.com>

driehle committed:
fixed PHP 8.1 type errors

Signed-off-by: Dennis Riehle <webmaster@riehle-web.com>

tobias-trozowski commented on phpcs.xml:
```diff
@@ -1,8 +1,18 @@
     <file>test</file>
+
+    <exclude-pattern>src/Int.php</exclude-pattern>
+    <exclude-pattern>src/Null.php</exclude-pattern>
```
we should follow the same structure here as in other components. 

@see https://github.com/laminas/repo-template/blob/main/phpcs.xml.dist

driehle replied:
Not sure what you mean, the linked file doesn't contain any exclude patterns.

tobias-trozowski replied:
sorry, my fault. i meant the overall structure of it.
e.g.
```xml
<?xml version="1.0"?>
<ruleset xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:noNamespaceSchemaLocation="./vendor/squizlabs/php_codesniffer/phpcs.xsd">

    <arg name="basepath" value="."/>
    <arg name="cache" value=".phpcs-cache"/>
    <arg name="colors"/>
    <arg name="extensions" value="php"/>
    <arg name="parallel" value="80"/>

    <!-- Show progress -->
    <arg value="p"/>

    <!-- Paths to check -->
    <file>src</file>
    <file>test</file>

    <exclude-pattern>src/Int.php</exclude-pattern>
    <exclude-pattern>src/Null.php</exclude-pattern>

    <!-- Include all rules from Laminas Coding Standard -->
    <rule ref="LaminasCodingStandard">
        <exclude name="SlevomatCodingStandard.TypeHints.ReturnTypeHint.MissingAnyTypeHint"/>
        <exclude name="SlevomatCodingStandard.TypeHints.ParameterTypeHint.MissingAnyTypeHint"/>
        <exclude name="SlevomatCodingStandard.TypeHints.PropertyTypeHint.MissingAnyTypeHint"/>
    </rule>
</ruleset>
```

driehle replied:
Fixed with 448ac52

tobias-trozowski commented on src/Compress/Gz.php:
```diff
@@ -152,19 +173,19 @@ public function compress($content)
 
         //check if there are null byte characters before doing a file_exists check
-        if (false === strpos($content, "\0") && file_exists($content)) {
+        if (null !== $content && false === strpos($content, "\0") && file_exists($content)) {
```
as unchanged lines cannot be commented i throw it here...
given `$content` is `null` it would raise an error further below:
```
} elseif ($mode === 'deflate') {
    $compressed = gzinflate($content);
} else {
    $compressed = gzuncompress($content);
}
```

this seems to be uncovered in tests (`DecompressTest` is testing this with bz2)

driehle replied:
Indeed, I added an explicit check for `null` values and extended tests in 73cd4bd

driehle replied:
See https://github.com/laminas/laminas-filter/pull/35#pullrequestreview-817519564 for follow up discussion on null values.

tobias-trozowski commented on src/DenyList.php:
```diff
@@ -40,7 +41,7 @@ public function setStrict($strict = true): void
      * Returns whether the in_array() call should be "strict" or not. See in_array docs.
      *
-     * @return boolean
+     * @return Boolean
```
```suggestion
     * @return bool
```

driehle replied:
Fixed with 448ac52

driehle committed:
updated PHPCs config and applied cs fixes

Signed-off-by: Dennis Riehle <webmaster@riehle-web.com>

driehle committed:
updated compress and decompress tests, explicitly test for null values in gz decompress

Signed-off-by: Dennis Riehle <webmaster@riehle-web.com>

driehle committed:
passing a passphrase null to openssl_pkey_get_private throws an error in PHP 7.3 and 7.4

Signed-off-by: Dennis Riehle <webmaster@riehle-web.com>

driehle commented on src/FilterPluginManager.php:
```diff
@@ -10,6 +12,29 @@
+use Zend\Filter\File\Rename;
+use Zend\Filter\File\RenameUpload;
+use Zend\Filter\File\UpperCase;
+use zend\filter\word\camelcasetodash;
+use zend\filter\word\camelcasetoseparator;
+use zend\filter\word\camelcasetounderscore;
+use zend\filter\word\dashtocamelcase;
+use zend\filter\word\dashtoseparator;
+use zend\filter\word\dashtounderscore;
+use zend\filter\word\separatortocamelcase;
+use zend\filter\word\separatortodash;
+use zend\filter\word\separatortoseparator;
+use zend\filter\word\underscoretocamelcase;
+use zend\filter\word\underscoretodash;
+use zend\filter\word\underscoretoseparator;
+use zend\filter\word\underscoretostudlycase;
```
It looks like PHPCs has broken the imports to the old Zend classes. Any ideas how to solve this?

driehle replied:
Restored aliases in da43764 and put them in a `@codingStandardsIgnoreStart/End` section, to prevent them from being changed by PHPCs again.

driehle committed:
restored aliases broken by PHPCs

Signed-off-by: Dennis Riehle <webmaster@riehle-web.com>

tobias-trozowski commented on test/DecompressTest.php:
```diff
@@ -73,56 +101,73 @@ public function testCompressToFile()
-            ]]
-        ];
+        foreach ($this->returnFilterType() as $parameter) {
+            yield [$parameter[0], new stdClass()];
```
previously available `null` was missing here
```suggestion
            yield [null];
            yield [$parameter[0], new stdClass()];
```

driehle replied:
It went further down into `testDecompressNullValueThrowsRuntimeException`. As that one is supposed to throw an exception, handling it in the test case above doesn't make sense to me.

tobias-trozowski replied:
in the previous version the test was expecting `null` as return when `null` was provided as argument, so this PR is actually introducing a new behaviour.

driehle replied:
Ok, I see. It looks like `bzdecompress(null)` actually works:

```
driehle:/laminas-filter$ php7.3 -r "var_dump(bzdecompress(null));"
string(0) ""
driehle:/laminas-filter$ php7.4 -r "var_dump(bzdecompress(null));"
string(0) ""
driehle:/laminas-filter$ php8.0 -r "var_dump(bzdecompress(null));"
string(0) ""
driehle:/laminas-filter$ php8.1 -r "var_dump(bzdecompress(null));"
string(0) ""
```

That is different from `gzuncompress(null)`, which has never worked and which throws a type incompatibility error with PHP 8.1:

https://3v4l.org/VJfrk

```
Deprecated: gzuncompress(): Passing null to parameter #1 ($data) of type string is deprecated in /in/VJfrk on line 3

Warning: gzuncompress(): data error in /in/VJfrk on line 3
bool(false)

Fatal error: Uncaught Error: Call to undefined function bzdecompress() in /in/VJfrk:6
Stack trace:
#0 {main}
  thrown in /in/VJfrk on line 6
```

So we would have to keep accepting `decompress(null)` for the `Bz2` filter, but allow it with the `Gz` filter... :-\

tobias-trozowski replied:
i would love to see all this fully typed and non-null... but yeah... need to keep accepting this.

driehle replied:
Well, there is an issue. Updating to laminas-coding-standard 2.0 has introduced `declare(strict_types=1)` in all files. This doesn't allow passing `null` to `bzuncompress()` anymore withouth getting a `TypeError`. I didn't recognize this at first, as my console commands posted above didn't include strict types. Here is the updated output:

```
driehle:/laminas-filter$ php7.3 -r "declare(strict_types=1);bzdecompress(null);"
PHP Fatal error:  Uncaught TypeError: bzdecompress() expects parameter 1 to be string, null given in Command line code:1
Stack trace:
#0 Command line code(1): bzdecompress(NULL)
#1 {main}
  thrown in Command line code on line 1
```

Output is the same for PHP 7.3, 7.4, 8.0 and 8.1.

I'll add something to cover null explicitly in the code.

tobias-trozowski replied:
the declare strict only kicks in for files which define it. if a have a `test.php` w/o `declare(strict_types=1)` there wont be a type error.

https://www.php.net/manual/en/language.types.declarations.php#language.types.declarations.strict

> Strict typing applies to function calls made from within the file with strict typing enabled, not to the functions declared within that file. If a file without strict typing enabled makes a call to a function that was defined in a file with strict typing, the caller's preference (coercive typing) will be respected, and the value will be coerced.

driehle replied:
You are right, but strict types are declared in our tests as well. I still think that we should cover this in the source files, so that it works for both users having strict types enabled or not enabled. See 289743d for a proposed solution.

driehle committed:
force PHPCs to keep compatibility with PHP 7.3

Signed-off-by: Dennis Riehle <webmaster@riehle-web.com>

driehle committed:
added explicit tests for decompressing null values with Bz2 and Gz

Signed-off-by: Dennis Riehle <webmaster@riehle-web.com>

driehle committed:
updated Psalm baseline and fixed cs issues

Signed-off-by: Dennis Riehle <webmaster@riehle-web.com>

tobias-trozowski commented on test/Compress/GzTest.php:
```diff
@@ -177,4 +186,15 @@ public function testGzToString()
+        $filter = new GzCompression();
+        $result = $filter->decompress(null);
+
+        $this->assertEmpty($result);
```
shouldn't be needed as decompress method is throwing exception with null arg

driehle replied:
Fixed with b9e3ef4

driehle committed:
simplified test and fixed type annotation

Signed-off-by: Dennis Riehle <webmaster@riehle-web.com>

driehle commented:
Is there something I can further help with?

weierophinney committed:
qa: treat deprecations as exceptions when testing

Helps us identify deprecated behavior early so we can make code forwards compatible.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
feat: bump minimum supported PHP version to 7.4

7.3 hits EOL in less than a week; no reason to support it in a new version.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: add one more item to Psalm baseline

Due to the definition of `FilterPluginManagerFactory::__invoke()`, and how it then calls the `FilterPluginManager` constructor.
No checks are needed at this time.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: add `#[ReturnTypeWillChange]` attribute to `FilterChain::count`

Fixes deprecation notice.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
